Client/Frontend/Home/RemoteTabsPanel.swift

```diff
@@ -13,160 +13,393 @@ import XCGLogger
     }
 
-    @objc private func SELrefresh() {
-        self.refreshControl?.beginRefreshing()
+    var tableViewDelegate: RemoteTabsPanelDataSource? {
+        didSet {
+            self.tableView.delegate = tableViewDelegate
+            self.tableView.dataSource = tableViewDelegate
+        }
+    }
 
-        self.profile.getClientsAndTabs().upon({ tabs in
-            if let tabs = tabs.successValue {
-                log.info("\(tabs.count) tabs fetched.")
-                self.clientAndTabs = tabs.filter { $0.tabs.count > 0 }
-
-                // Maybe show a background view.
-                let tableView = self.tableView
-                if let clientAndTabs = self.clientAndTabs where clientAndTabs.count > 0 {
-                    tableView.backgroundView = nil
-                    // Show dividing lines.
-                    tableView.separatorStyle = UITableViewCellSeparatorStyle.SingleLine
-                } else {
-                    // TODO: Bug 1144760 - Populate background view with UX-approved content.
-                    tableView.backgroundView = UIView()
-                    tableView.backgroundView?.frame = tableView.frame
-                    tableView.backgroundView?.backgroundColor = UIColor.redColor()
+    func refresh() {
+        tableView.scrollEnabled = false
+        tableView.allowsSelection = false
+        tableView.tableFooterView = UIView(frame: CGRectZero)
+
+        // Short circuit if the user is not logged in
+        if profile.getAccount() == nil {
+            self.tableViewDelegate = RemoteTabsPanelErrorDataSource(homePanel: self, error: .NotLoggedIn)
+            self.tableView.reloadData()
+            return
+        }
 
-                    // Hide dividing lines.
-                    tableView.separatorStyle = UITableViewCellSeparatorStyle.None
+        // Otherwise, fetch the tabs cloud
+        self.profile.getClientsAndTabs().upon({ result in
+            self.refreshControl?.endRefreshing()
+
+            if let clientAndTabs = result.successValue {
+                if clientAndTabs.count == 0 {
+                    self.tableViewDelegate = RemoteTabsPanelErrorDataSource(homePanel: self, error: .NoClients)
+                } else {
+                    let nonEmptyClientAndTabs = clientAndTabs.filter { $0.tabs.count > 0 }
+                    if nonEmptyClientAndTabs.count == 0 {
+                        self.tableViewDelegate = RemoteTabsPanelErrorDataSource(homePanel: self, error: .NoTabs)
+                    } else {
+                        self.tableViewDelegate = RemoteTabsPanelClientAndTabsDataSource(homePanel: self, clientAndTabs: nonEmptyClientAndTabs)
+                        self.tableView.allowsSelection = true
+                    }
                 }
-                tableView.reloadData()
             } else {
-                log.error("Failed to fetch tabs.")
+                self.tableViewDelegate = RemoteTabsPanelErrorDataSource(homePanel: self, error: .FailedToSync)
```

Comment:
Follow-up: don't enter an error state here, show a bar at the top and show your old tabs from the DB.